pandas/core/reshape/reshape.py

```diff
@@ -741,13 +741,11 @@ def _convert_level_number(level_num: int, columns):
             chunk.columns = level_vals_nan.take(chunk.columns.codes[-1])
             value_slice = chunk.reindex(columns=level_vals_used).values
         else:
-            if frame._is_homogeneous_type and is_extension_array_dtype(
-                frame.dtypes.iloc[0]
-            ):
+            subset = this[this.columns[loc]]
```

Comment:
can we avoid making a copy here?

Reply:
yes that's a great point.

It's easy to do for the is_extension_array_dtype test, but not immediately obvious how to do best for the homogeneity test.

I can think of three ways
- take an empty slice and do the homogeneity test on that, sth like subset_is_homogeneous_type = this.iloc[:0][this.columns[loc]]._is_homogeneous_type
- subselect the dtypes, build an empty frame, like pd.DataFrame(dtype=this.dtypes[this.columns[loc]])._is_homogeneous_type
- inline the logic from _is_homogeneous_type 

I'd lean towards for the first.

Is there a preferred way here?

Reply:
> I'd lean towards for the first.

this seems reasonable.  make sure to put a comment about why you're doing it

Reply:
updated the code to avoid the copy.

Played around a bit to see if the logic can be refactored. There's a subtle interplay there between the dtypes of the frame / subset. Plus the fact that "heterogeneous" and "homogeneous" are not the obvious things when extension arrays are involved.

Given that this piece of code will be wiped out anyways once the 2d-extension arrays come around, I've settled for the minimal change now.

Please let me know if this looks OK to you